### PR TITLE
[v6r22] DataRecoveryAgent from iLCDirac

### PR DESCRIPTION
**Pull request by andresailer**
BEGINRELEASENOTES

*TS
NEW: DataRecoveryAgent to perform and apply consistency checks for transformations
NEW: dirac-transformation-recover-data : script to manually run consistency checks on individual transformations, for debugging and testing of the DataRecoveryAgent

ENDRELEASENOTES

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ numpy>=1.10.1
 pexpect>=4.0.1
 psutil>=4.2.0
 Pygments>=1.5
+parameterized
```
Comment (chaen) [inline requirements.txt]: What's the difference with the `parametrize` included in `pytest`?

Comment (andresailer) [reply requirements.txt]: It also works for unittest based tests

Comment (andresailer) [reply requirements.txt]: I did not reallise until after making the pr that I had this dependency in the tests. I can of course rebase on any other branch.

File: TransformationSystem/ConfigTemplate.cfg [hunk 2]
```diff
@@ -180,5 +180,24 @@ Agents
...
+    MailTo =
+    MailFrom =
+    # List of TransformationIDs that will not be treated
+    TransformationsToIgnore =
+    # List of Transformation Statuses to treat
+    TransformationStatus = Active, Completing
+    # List of transformations that do not have input data
+    TransformationsNoInput = MCGeneration
+    # List of transformations that do have input data
+    TransformationsWithInput = MCReconstruction, MCSimulation, MCReconstruction_Overlay
```
Comment (fstagni) [inline TransformationSystem/ConfigTemplate.cfg]: Other TS agents use these locations:
```
    transformationTypes = Operations().getValue('Transformations/DataProcessing', [])
    extendableTTypes = Operations().getValue('Transformations/ExtendableTransfTypes', ['MCSimulation'])
```
Which might not cover fully your case here, but nevertheless if more options need to be added should be in `/Operations/Transformations`.

File: TransformationSystem/Agent/DataRecoveryAgent.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+from DIRAC.Core.Utilities.List import breakListIntoChunks
+from DIRAC.WorkloadManagementSystem.Client.JobMonitoringClient import JobMonitoringClient
+from DIRAC.Resources.Catalog.FileCatalogClient import FileCatalogClient
+from DIRAC.TransformationSystem.Client.TransformationClient import TransformationClient
+from DIRAC.RequestManagementSystem.Client.ReqClient import ReqClient
+from DIRAC.FrameworkSystem.Client.NotificationClient import NotificationClient
+
+from DIRAC.TransformationSystem.Utilities.TransformationInfo import TransformationInfo
+from DIRAC.TransformationSystem.Utilities.JobInfo import TaskInfoException
+from DIRAC.Interfaces.API.Dirac import Dirac
```
Comment (fstagni) [inline TransformationSystem/Agent/DataRecoveryAgent.py]: Re-order imports

File: TransformationSystem/ConfigTemplate.cfg (hunk 2 above)
Comment (andresailer) [reply TransformationSystem/ConfigTemplate.cfg]: It is safe to assume that `ExtendableTransfTypes` contains transformations without input data, right? Thus `DataProcessing - ExtendableTransfTypes` gives me transformationsWithInput

File: TransformationSystem/Agent/DataRecoveryAgent.py (hunk 3 above)
Comment (andresailer) [reply TransformationSystem/Agent/DataRecoveryAgent.py]: Sorted imports

File: TransformationSystem/ConfigTemplate.cfg (hunk 2 above)
Comment (fstagni) [reply TransformationSystem/ConfigTemplate.cfg]: I would not think it's a 100% safe assumption. It works in LHCb and IIUC in CLIC but not sure about other VOs.

Comment (andresailer) [reply TransformationSystem/ConfigTemplate.cfg]: Though can I use those assumption for the default values?
Things can be overwritten in the agent configuration.

File: TransformationSystem/Agent/DataRecoveryAgent.py [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+  files to obtain consistent state for jobs, tasks, input and output
+  files
+
+- Send email about performed actions
+
+Requirements/Assumptions:
+
+- JobParameters:
+
+  - ProductionOutputData: with the list semi-colon separated list of expected output files, stored as a Job Parameter
```
Comment (andresailer) [inline TransformationSystem/Agent/DataRecoveryAgent.py]: This is the last use of "Production" in the agent and related files. Should I add an option to change the parameter under which the OutputData files can be found?

**Comment by andresailer** [comment]
Review please!

File: TransformationSystem/Agent/DataRecoveryAgent.py (hunk 4 above)
Comment (fstagni) [reply TransformationSystem/Agent/DataRecoveryAgent.py]: I am afraid that this is used in so many places that changing it would risky...

**Pull request merged**